### PR TITLE
Replace NullArrayType with ArrayType<UndefinedType>

### PR DESCRIPTION
**Pull request by romseygeek**
Adding a special ArrayType to just handle nulls means lots of extra special
case handling in various places.  Making some small changes to UndefinedType
so that it can be used here instead is a lot simpler.

Note that this moves some tests from IndexerTests (which does not handle dynamic
updates well) into SQLTypeMappingTest, and adds tests for the error cases that
Minsoo found.

### CODE REVIEW

File: server/src/main/java/io/crate/types/DataTypes.java [hunk 1]
```diff
@@ -666,4 +659,8 @@ public static DataType<?> fromId(Integer id) {
             .findFirst()
             .orElse(DataTypes.UNDEFINED);
     }
+
+    public static boolean isArrayOfNulls(DataType<?> type) {
+        return type instanceof ArrayType<?> at && at.innerType() == UndefinedType.INSTANCE;
+    }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/types/DataTypes.java:665]:
```suggestion

    public static boolean isArrayOf(DataType<?> type, DataType<?> innerType) {
        return type instanceof ArrayType<?> at && at.innerType().id() == innerType.id();
    }
```

Not sure if worth it, but would make it a bit more generic.

**Review by mfussenegger — APPROVED** [review]

File: server/src/test/java/io/crate/execution/dml/IndexerTest.java [hunk 2]
```diff
@@ -1208,44 +1177,6 @@ public void test_empty_arrays_together_with_another_field_added_as_new_cols_with
...
-        );
-        var item = item(List.of());
-        List<Reference> newColumns = indexer.collectSchemaUpdates(item);
-        assertThat(newColumns).hasSize(1);
-
-        ParsedDocument doc = indexer.index(item);
-        assertThat(doc.source().utf8ToString()).isEqualToIgnoringWhitespace(
-                """
-                {"_u_empty_arr":[]}
-                """
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/execution/dml/IndexerTest.java:1235]: Super minor: let's update javadoc of `DocTableInfo.lookupNameBySourceKey`
> *  <li>Empty arrays, or arrays with only null values</li> 

is not true anymore

File: server/src/test/java/io/crate/integrationtests/SQLTypeMappingTest.java [hunk 3]
```diff
@@ -463,6 +463,40 @@ public void test_dynamic_null_array_overridden_to_integer_becomes_null() throws
...
+                {a=1, n1=[], n2=[null, null], o={x=[], y=[null]}}""",
+            """
+                {a=2, n1=[1], n2=[2], o={x=[1], y=[1]}}""");
+    }
+
+    public void test_dynamic_nested_array_of_null_upcast() {
+        execute("create table t (a int) with (column_policy= 'dynamic')");
+        Asserts.assertSQLError(() -> execute("insert into t(a,b) values (1, [[]])"))
+                .hasMessageContaining("Dynamic nested arrays are not supported");
+    }
```
Comment (jeeminso) [inline server/src/test/java/io/crate/integrationtests/SQLTypeMappingTest.java:486]: There seems to be a way to work around this to created dynamic nested arrays:
```
cr> create table t (a int) with (column_policy= 'dynamic');
CREATE OK, 1 row affected (1.667 sec)
cr> insert into t(a,b) values (1, []);
INSERT OK, 1 row affected (0.161 sec)
cr> insert into t(a,b) values (1, [[]]);
INSERT OK, 1 row affected (0.162 sec)
```

Comment (romseygeek) [reply server/src/test/java/io/crate/integrationtests/SQLTypeMappingTest.java:486]: Sneaky.  Good catch, I will see if I can detect this case and throw.

**Review by BaurzhanSakhariev — APPROVED** [review]
Thanks!

**Review by jeeminso — APPROVED** [review]
LGTM 👍

**Comment by romseygeek** [comment]
@Mergifyio squash title+body

**Pull request merged**